FreeRTOS-Plus/Source/FreeRTOS-IoT-Libraries-LTS-Beta2/c_sdk/platform/freertos/transport/src/tls_freertos.c

```diff
@@ -0,0 +1,604 @@
+        returnStatus = TLS_TRANSPORT_INVALID_PARAMETER;
+    }
+
+    /* Create a FreeRTOS+TCP socket. */
+    if( returnStatus == TLS_TRANSPORT_SUCCESS )
+    {
+        socketStatus = Sockets_Connect( &( pNetworkContext->tcpSocket ),
+                                        pHostName,
+                                        port,
+                                        receiveTimeoutMs,
+                                        sendTimeoutMs );
+
+        if( socketStatus != 0 )
+        {
+            LogError( ( "Failed to connect to %s with error %d.",
+                        pHostName,
+                        socketStatus ) );
+            returnStatus = TLS_TRANSPORT_CONNECT_FAILURE;
+        }
+    }
+
+    /* Initialize mbedtls. */
+    if( returnStatus == TLS_TRANSPORT_SUCCESS )
+    {
+        returnStatus = initMbedtls();
+    }
+
+    /* Perform TLS handshake. */
+    if( returnStatus == TLS_TRANSPORT_SUCCESS )
+    {
+        returnStatus = tlsSetup( pNetworkContext, pHostName, pNetworkCredentials );
+    }
+
+    /* Clean up on failure. */
+    if( returnStatus != TLS_TRANSPORT_SUCCESS )
+    {
+        if( pNetworkContext->tcpSocket != FREERTOS_INVALID_SOCKET )
+        {
+            FreeRTOS_closesocket( pNetworkContext->tcpSocket );
+        }
+    }
+    else
+    {
+        LogInfo( ( "(Network connection %p) Connection to %s established.",
+                   pNetworkContext,
+                   pHostName ) );
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+void TLS_FreeRTOS_Disconnect( NetworkContext_t * pNetworkContext )
+{
+    BaseType_t socketStatus = 0;
+
+    socketStatus = ( BaseType_t ) mbedtls_ssl_close_notify( &( pNetworkContext->sslContext.context ) );
+
+    /* Ignore the WANT_READ and WANT_WRITE return values. */
+    if( ( socketStatus != ( BaseType_t ) MBEDTLS_ERR_SSL_WANT_READ ) &&
+        ( socketStatus != ( BaseType_t ) MBEDTLS_ERR_SSL_WANT_WRITE ) )
+    {
+        if( socketStatus == 0 )
+        {
+            LogInfo( ( "(Network connection %p) TLS close-notify sent.",
+                       pNetworkContext ) );
+        }
+        else
+        {
+            LogError( ( "(Network connection %p) Failed to send TLS close-notify, error %d.",
+                        pNetworkContext,
+                        socketStatus ) );
+        }
+    }
+
+    /* Call socket shutdown function to close connection. */
+    Sockets_Disconnect( pNetworkContext->tcpSocket );
+
+    /* Free mbed TLS contexts. */
+    sslContextFree( &( pNetworkContext->sslContext ) );
+}
+
+/*-----------------------------------------------------------*/
+
+int32_t TLS_FreeRTOS_recv( NetworkContext_t * pNetworkContext,
+                           void * pBuffer,
+                           size_t bytesToRecv )
+{
+    int32_t socketStatus = 0;
+
+    socketStatus = ( int32_t ) mbedtls_ssl_read( &( pNetworkContext->sslContext.context ),
```

Comment:
Should the status variable be of `BaseType_t` type (the `TLS_FreeRTOS_Disconnect` uses that type)?

Reply:
no the function signature of a transport read is int32_t.